contracts/staking/fuse/utils/StakingRewardsPerEpoch.sol

```diff
@@ -0,0 +1,201 @@
+	event Recovered(address token, uint256 amount);
+
+	bytes32 public constant GUARDIAN_ROLE = keccak256("GUARDIAN_ROLE");
+	uint256 public constant PRECISION = 1e18;
+
+	IERC20 public rewardsToken;
+	IERC20 public stakingToken;
+
+	mapping(address => StakeInfo) public stakersInfo;
+
+	uint256 public lastEpochIndex;
+	uint256 public pendingStakes;
+	uint256[] public rewardsPerTokenAt;
+
+	uint256 internal _totalSupply;
+
+	modifier updateReward(address account) {
+		_updateReward(account);
+		_;
+	}
+
+	constructor(address _rewardsToken, address _stakingToken) {
+		_setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+		_setupRole(GUARDIAN_ROLE, msg.sender);
+		rewardsToken = IERC20(_rewardsToken);
+		stakingToken = IERC20(_stakingToken);
+	}
+
+	function totalSupply() external view returns (uint256) {
+		return _totalSupply;
+	}
+
+	function balanceOf(address account) external view returns (uint256) {
+		return stakersInfo[account].balance;
+	}
+
+	function stake(uint256 amount)
+		external
+		payable
+		nonReentrant
+		whenNotPaused
+		updateReward(msg.sender)
+	{
+		require(amount > 0, "Cannot stake 0");
+		_stake(msg.sender, amount);
+	}
+
+	function withdraw(uint256 _amount)
+		public
+		nonReentrant
+		updateReward(msg.sender)
+	{
+		require(_amount > 0, "Cannot withdraw 0");
+		_withdraw(msg.sender, _amount);
+	}
+
+	function getReward() public nonReentrant updateReward(msg.sender) {
+		_getReward(msg.sender);
+	}
+
+	function exit() external {
+		withdraw(stakersInfo[msg.sender].balance);
+		getReward();
+	}
+
+	function rewardPerToken() public view returns (uint256) {
+		if (_totalSupply == 0) {
+			return rewardPerTokenAt[0];
+		}
+		return
+			rewardPerTokenStored +
+			((lastTimeRewardApplicable() - lastUpdateTime) * rewardRate * PRECISION) /
+			_totalSupply;
+	}
+
+	function _getRewardPerTokenPerUser(address _account)
+		internal
+		view
+		returns (uint256)
+	{
+		return
+			rewardsPerTokenAt[lastEpochIndex] -
+			rewardsPerTokenAt[stakersInfo[_account].indexOfLastEpochStaked + 1];
+	}
+
+	function earned(address account) public view  returns (uint256) {
+		return
+			(stakersInfo[account].balance *
+				(rewardPerToken() -
+					stakersInfo[account].rewardPerTokenPaid)) /
+			PRECISION +
+			stakersInfo[account].reward;
+	}
+
+	function _updateReward(address _account) internal virtual {
+		lastUpdateTime = lastTimeRewardApplicable();
+		if (_account != address(0)) {
+			_addPendingStakesToBalanceOnTimeUpdate(_account);
+			stakersInfo[_account].reward = earned(_account);
+			stakersInfo[_account]
+					.rewardPerTokenPaid = _getRewardPerTokenPerUser(_account);
+		}
+	}
+
+	function _addPendingStakesToBalanceOnTimeUpdate(address _account) internal {
+		if (
+					stakersInfo[_account].indexOfLastEpochStaked !=
+					lastEpochIndex
+		) {
+				stakersInfo[_account].balance += stakersInfo[_account]
+						.pendingStake;
+				stakersInfo[_account].pendingStake = 0;
+		}
+	}
+
+	function _updatePeriodAndRate(uint256 reward) internal virtual {
```

Comment:
i dont understand why we have this function, is that logic defined anywhere in the requirements?

Reply:
That was the vanilla Synthetix notifyRewardAmount function. We have to use this function to calculate reward rate, that will be multiplied on duration of time interval between two collect UBI's. The result of multiplication is the exact amount of reward that must be distributed to the users.